File: Part 7 - Natural Language Processing/Natural Language Processing/NLP_Swats.R
#Natural Language Processing

#Importing the dataset
#Ignore the quotes in the text
#To prevent the reviews from being identified as factors
dataset_original = read.delim('Restaurant_Reviews.tsv', quote='', stringsAsFactors = FALSE)

#Cleaning the texts
#install.packages('tm')
#install.packages('SnowballC')
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset_original$Review))

#Converting all words to lowercase to have one unique version of the word
corpus = tm_map(corpus, content_transformer(tolower))
#Removing the numbers from the reviews
corpus = tm_map(corpus, removeNumbers)
#Remove punctuation to avoid having a separate column for each punctuation mark
corpus = tm_map(corpus, removePunctuation)
#Remove irrelavant words
corpus = tm_map(corpus, removeWords, stopwords())
#Stemming Words
corpus = tm_map(corpus, stemDocument)
#Removing extra white spaces
corpus = tm_map(corpus, stripWhitespace)

#Creating Bag of Words Model
dtm = DocumentTermMatrix(corpus)
#Filter words that appear only once and not frequently in the review
dtm = removeSparseTerms(dtm, 0.999)
dataset = as.data.frame(as.matrix(dtm))
dataset$Liked = dataset_original$Liked

#You need to factor 0,1 because otheriwse R would treat them as numeric. We want R to interpret it as characters.
# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

#Random Forest Classififer
# Splitting the dataset into the Training set and Test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Liked, SplitRatio = 0.80)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Fitting Random Forest Classification to the Training set
# install.packages('randomForest')
library(randomForest)
set.seed(123)
classifier = randomForest(x = training_set[-692],
                          y = training_set$Liked,
                          ntree = 10)

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set[-692])

# Making the Confusion Matrix
cm = table(test_set[, 692], y_pred)

File: Part 7 - Natural Language Processing/Natural Language Processing/NLP_Swats.py
#Natural Language Processing

#Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#To ignore the double quotes, we add quoting = 3, 3 is special code for ignoring quotes
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

#Cleaning the text to make it ready for future machine learning algorithm
#Compulsory step for NLP

#Cleaning the texts
import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
#nltk.download('stopwords')
#Corpus - Collection of texts
corpus = []
for i in range(0, 1000):
    review = re.sub('[^a-zA-Z]',' ',dataset['Review'][i])
    review = review.lower()
    review = review.split()
    #Use set rather than review - set is faster than review
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

#Create bag of words model    
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 1]

#Applying Naive Bayes Algorithm
# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

################################################################################################################################
# Fitting Naive Bayes to the Training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

accuracy = (cm[0][0] + cm[1][1])/(cm[0][0] + cm[1][1] + cm[0][1] + cm[1][0])
precision = cm[1][1]/(cm[1][1] + cm[0][1])
recall = cm[1][1]/ (cm[1][1] + cm[1][0])
f1_score = (2 * precision * recall) / (precision + recall)

###############################################################################################################################
#Decision Tree Classifier

# Fitting Decision Tree classifier to the Training set
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit (X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

accuracy = (cm[0][0] + cm[1][1])/(cm[0][0] + cm[1][1] + cm[0][1] + cm[1][0])
precision = cm[1][1]/(cm[1][1] + cm[0][1])
recall = cm[1][1]/ (cm[1][1] + cm[1][0])
f1_score = (2 * precision * recall) / (precision + recall)

################################################################################################################################
#Logistic Regression
# Fitting Logistic Regression to the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

accuracy = (cm[0][0] + cm[1][1])/(cm[0][0] + cm[1][1] + cm[0][1] + cm[1][0])
precision = cm[1][1]/(cm[1][1] + cm[0][1])
recall = cm[1][1]/ (cm[1][1] + cm[1][0])
f1_score = (2 * precision * recall) / (precision + recall)

################################################################################################################################
#Bernoulli Naive Bayes
# Fitting Naive Bayes to the Training set
from sklearn.naive_bayes import BernoulliNB
classifier = BernoulliNB()
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

accuracy = (cm[0][0] + cm[1][1])/(cm[0][0] + cm[1][1] + cm[0][1] + cm[1][0])
precision = cm[1][1]/(cm[1][1] + cm[0][1])
recall = cm[1][1]/ (cm[1][1] + cm[1][0])
f1_score = (2 * precision * recall) / (precision + recall)